### PR TITLE
DEX-888 (etc): Twitter / Intelligent Agent

### PR DESCRIPTION
**Pull request by naknomum**
## Pull Request Overview

- Adds support for TwitterBot _Intelligent Agent_ and framework for extending to other sources, such as Instagram, Gopher, etc.
- New base class `IntelligentAgentContent` (and current subclass `TwitterTweet`) is persisted to db per incoming object (e.g. tweet)
- Introduces SiteSettings for above -- auto-generated from corresponding Agent extension
- Implements (but does not use; see below) `linked_accounts` on Users

---

**Review Notes**
- **General overview:** site admin sets appropriate Twitter-provided credentials (and enable TwitterBot); periodic polling of Twitter Account for tweets; if tweets contain enough information - they are processed as submissions; if not, a reply-tweet to the user is sent.
- Tweet must have: image(s), hashtag of species,  region/location id, time/date.  There is some flexibility and fuzziness in how this is matched; but improvements are surely needed and forthcoming as data comes in.
- If the tweet-author can be found via a linked account, they are assigned as owner of generated Encounters/data.  If not, the data is owned by _public user_.
- Generally, incoming data follows "standard" submission route.  Content is manipulated into an AssetGroup (with Assets) along with AssetGroupSighting (metadata).  If applicable, this is auto-committed to a full-fledged Sighting/Encounter (and moves on to identification).  As such, `IntelligentAgentContent` is linked via `content.asset_group`.
- Twitter connectivity via [Tweepy](https://docs.tweepy.org/en/stable/), favoring the V2 "Client" implementation.
- PR introduces using `gettext` for [i18n of text](https://docs.python.org/3/library/gettext.html) on the backend.  This will allow supporting multiple languages for messages used with Intelligent Agents.  There is a `locales/` directory for this purpose and text is wrapped in the `_(text)` method as required.  _This enhancement can be explored more in the future and is just the basic minimum for now._

**Notes on _linked accounts_**

Details are given here.  :warning: However, due to time constraints, this _is not used_ for the current twitter implementation; using instead a less secure but simple `twitter_username` property on the User.

- `user.linked_accounts` is a json object with a generic `{ service_key: data, ... }` structure.  It is accessed via `user.link_account(key, data)` and (classmethod) `user = User.find_by_linked_account(key, value)`
- Further, a new api `/api/v1/users/social_callback/SERVICE` allows for a callback to use proper channels to authenticate services to link accounts.  This is implemented for `/twitter` based on the ["Login with Twitter"](https://developer.twitter.com/en/docs/authentication/guides/log-in-with-twitter) protocol.
- Likewise, in order to _initiate_ authenticated protocols such as above, `/api/v1/users/social_auth_redirect/SERVICE` api is provided, which will redirect the user to the appropriate location on the service to authenticate their account (as well as prep the backend accordingly, using the site's credentials).  Once the user agrees, they are redirected back to the codex site (hard-coded as their _settings page_ presently) and are then officially linked via `user.linked_accounts`.

**Model Updates**
Not comprehensive, just an overview.

`IntelligentAgent` (base class; `TwitterBot` implemented subclass)
- `collect()` (get content)
- `test_setup()` (verify settings are correct)
- `is_enabled()` and `is_ready()`
- misc. other utility functions

`IntelligentAgentContent` (base; `TwitterTweet` implemented)
- `guid` (etc)
- `state` (currently: `intake`, `rejected`, `active`, `active_detection`, `active_identification`, `complete`, `error`)
- `raw_content` (json)
- `source` (json)
- `owner_guid` (and `owner`)
- `asset_group_guid` (`asset_group`)
- `data` (useful data _derived_ from content and metadata)
- misc methods to get this information in various useful formats
- `respond_to()` - send a reply content in reference to this content

**REST API Updates**

In addition to the account-linking endpoints discussed above, an endpoint meant to be expanded as needed is introduced: `/api/v1/site-settings/test/`.  For the Intelligent Agent functionality, the sub-endpoint `/api/v1/site-settings/test/intelligent_agent_NAME` is used.  Based on the results of `IntelligentAgent.test_setup()` either a `400` is returned (settings are not working; some details may be given in json response) or `200` success, if good.  This endpoint is only accessible to users with _site-setting WRITE_ permissions.  Example:

```
[GET] /api/v1/site-settings/test/intelligent_agent_twitterbot
200
{
    "success": true,
    "message": "Success: Twitter username is 'foo', name is 'Fubar'",
    "username": "foo",
    "name": "Fubar"
}
```

**Extra bits**
- To persist small bits of data (e.g. last tweet collected) and _not_ have to store in db table, introduced the utility methods `app.utils.set_persisted_value(key, value)` and `get_persisted_value(key)` which use _Redis_ to store the data.
- Further, these are wrapped, via `IntelligentAgent` class as class methods (e.g. `IntelligentAgent.get_persisted_value()`) which automatically prefix the _key_ with a corresponding string, such as `intelligent_agent_twitterbot_` 
- Using the package [FuzzyWuzzy](https://pypi.org/project/fuzzywuzzy/), a utility `app.utils.fuzzy_match()` is introduced for general usage.  It is employed by both taxonomy and region matching in IntelligentAgent content.
- Similar to the `Regions` class (under site_settings), there is now a `Taxonomy` class.  It has some general-purpose utilities for parsing and finding taxonomies, and makes use of the site configuration ("species for this codex") values.

**Celery background tasks**
- Main _periodic_ task calls `TwitterBot.collect()` at regular `polling_interval` (set via site-settings; but currently a bug does not allow it to be dynamically changed)
- When a tweet/content makes it to _detection_ and/or _identification_, a task fires up to wait for results.  It will timeout (and let user know that nothing happened) if applicable.

### CODE REVIEW

**Review by colinwkingen — COMMENTED** [review]
Comments are all higher level, code itself looks good. Will chat about the module/submodule structuring.

File: app/extensions/intelligent_agent/models.py [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+log = logging.getLogger(__name__)  # pylint: disable=invalid-name
+_ = gettext.gettext
+
+
+class TwitterBot(IntelligentAgent):
+    """
+    TwitterBot intelligent Agent
+    """
+
+    # note, `api` is for V1 Twitter API and `client` if for V2; we obtain both, but probably should
```
[the remark this answers is not in the document]
Comment (naknomum) [reply app/extensions/intelligent_agent/models.py:28]: yes:  currently you cannot send images as part of a tweet with V2.  they are "working on it".  so i left V1 `api` around.  there may be others, but that was the one i noted.

File: app/extensions/intelligent_agent/models.py [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+            or self.get_persisted_value('twitter_outgoing_disabled') == 'true'
+        ):
+            log.warning('create_tweet_direct(): OUTGOING DISABLED')
+            return
+        tweet = self.client.create_tweet(text=text, in_reply_to_tweet_id=in_reply_to)
+        log.debug(f'create_tweet_direct(): success tweeting {tweet}')
+        return tweet
+
+    # preferred usage (vs create_tweet_direct()) as it will throttle outgoing rate to
+    #   hopefully keep twitter guards happy
```
[the remark this answers is not in the document]
Comment (naknomum) [reply app/extensions/intelligent_agent/models.py:145]: there are definitely **so many** limitations on the twitter api.  queries, searches, posting of actual tweets, etc.  way too many for me to list, let alone undersand.  :smile:   definitely in old-world (with the older twitter api) we had times when (due to our own logic mostly) we would just kinda barf out a whole string of outgoing tweets and trigger a (temporary) lockout.  hence me trying to throttle just how fast they go out.   if ever we are busy enough to need to alter the values i have, we can do so -- but carefully.

File: app/extensions/intelligent_agent/__init__.py [hunk 3]
```diff
@@ -0,0 +1,605 @@
+# -*- coding: utf-8 -*-
+# pylint: disable=no-self-use
+"""
+Base Intelligent Agent
+
+"""
```
[the remark this answers is not in the document]
Comment (drewblount) [reply app/extensions/intelligent_agent/__init__.py:6]: Yeah, my gut would be to put all these classes in a single models.py file and have a relatively short __init__.py, more in the style of our app modules than the elasticsearch extension..

**Comment by colinwkingen** [comment]
TGML (Too good. Me looked.)

**Pull request merged**